compiler/fir/resolve/src/org/jetbrains/kotlin/fir/resolve/transformers/InvocationKindTransformer.kt

```diff
@@ -6,98 +6,49 @@
+    val argumentMapping = calleeReference.candidate.argumentMapping ?: return
+    val function = calleeReference.candidateSymbol.fir as? FirSimpleFunction ?: return
+    val isInline = function.isInline
+
+    val byParameter = mutableMapOf<FirValueParameter, EventOccurrencesRange>()
+    function.contractDescription.effects?.forEach { fir ->
+        val effect = fir.effect as? ConeCallsEffectDeclaration ?: return@forEach
+        // TODO: Support callsInPlace contracts on receivers
+        val valueParameter = function.valueParameters.getOrNull(effect.valueParameterReference.parameterIndex) ?: return@forEach
+        byParameter[valueParameter] = effect.kind
     }
-
-    override fun transformFunctionCall(functionCall: FirFunctionCall, data: Any?): FirStatement {
-        val calleeReference = functionCall.calleeReference as? FirNamedReferenceWithCandidate ?: return functionCall
-        val argumentMapping = calleeReference.candidate.argumentMapping ?: return functionCall
-        val function = calleeReference.candidateSymbol.fir as? FirSimpleFunction ?: return functionCall
-
-        val callsEffects = function.contractDescription.effects
-            ?.map { it.effect }
-            ?.filterIsInstance<ConeCallsEffectDeclaration>() ?: emptyList()
-
-        val isInline = function.isInline
-        if (callsEffects.isEmpty() && !isInline) {
-            return functionCall
-        }
-
-        val reversedArgumentMapping = argumentMapping.entries.map { (argument, parameter) ->
-            parameter to argument
-        }.toMap()
-
-        val invocationKindMapping = mutableMapOf<FirExpression, EventOccurrencesRange>()
-        for (effect in callsEffects) {
-            // TODO: Support callsInPlace contracts on receivers
-            val valueParameter = function.valueParameters.getOrNull(effect.valueParameterReference.parameterIndex) ?: continue
-            val argument = reversedArgumentMapping[valueParameter] ?: continue
-            invocationKindMapping[argument] = effect.kind
-        }
-        if (isInline) {
-            for (argument in functionCall.arguments) {
-                invocationKindMapping.putIfAbsent(argument, EventOccurrencesRange.UNKNOWN)
-            }
-        }
-        if (invocationKindMapping.isEmpty()) {
-            return functionCall
-        }
-        functionCall.argumentList.transformArguments(ArgumentsTransformer, invocationKindMapping to null)
-        return functionCall
+    if (byParameter.isEmpty() && !isInline) return
+
+    for ((argument, parameter) in argumentMapping) {
+        val lambda = argument.unwrapAnonymousFunctionExpression() ?: continue
+        val kind = byParameter[parameter] ?: EventOccurrencesRange.UNKNOWN.takeIf {
+            // Inline functional parameters have to be called in-place; that's the only permitted operation on them.
+            isInline && !parameter.isNoinline && !parameter.isCrossinline &&
```

Comment:
Actually you should check only for `noinline` modifier here. `crossinline` keeps the lambda inline (so you are required to call it in place). The only difference is that `crossinline` prohibit non-local return from lambda body

Reply:
Wow, TIL something new. Thank you